pyuvdata/uvdata/corr_fits_src/corr_fits.pyx

```diff
@@ -121,3 +124,124 @@ cpdef numpy.ndarray[ndim=1, dtype=numpy.float64_t] get_cable_len_diffs(
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+cdef numpy.ndarray[ndim=2, dtype=numpy.float64_t] _compute_khat(
+  numpy.float64_t[:, ::1] x,
+  numpy.ndarray[ndim=1, dtype=numpy.float64_t] sig1,
+  numpy.ndarray[ndim=1, dtype=numpy.float64_t] sig2,
+):
+
+  cdef numpy.ndarray[ndim=2, dtype=numpy.float64_t] khat = np.zeros((x.shape[0], x.shape[1]), dtype=np.float64)
+
+  cdef numpy.float64_t[:, ::1] _khat = khat
+
+  ind1 = np.arange(0.5, 7.5, 1, dtype=np.float64).reshape(7, 1)
+  cdef numpy.float64_t[:, ::1]  j_ind = ind1 / sig1
+  cdef numpy.float64_t[:, ::1]  k_ind = ind1 / sig2
+
+  cdef Py_ssize_t i, j, k, l
+
+  if x.size > 800:  # pragma: nocover
+    with nogil, parallel():
+      for j in prange(x.shape[1]):
+        for i in range(x.shape[0]):
+          for l in range(k_ind.shape[0]):
+            for k in range(j_ind.shape[0]):
+              _khat[i, j] += (
+                1./ (pi * sqrt( 1 - x[i, j] ** 2)) * (
+                  exp(-1. / (2 * (1 - x[i,j] ** 2)) * (j_ind[k, j] ** 2 + k_ind[l, j] ** 2 - 2 * x[i, j] * j_ind[k, j] * k_ind[l,j]))
+                  + exp(-1. / (2 * (1 - x[i,j] ** 2)) * (j_ind[k, j] ** 2 + k_ind[l, j] ** 2 + 2 * x[i, j] * j_ind[k, j] * k_ind[l,j]))
+                )
+              )
+  else:
+    for i in range(x.shape[0]):
+      for j in range(x.shape[1]):
+        for l in range(k_ind.shape[0]):
+          for k in range(j_ind.shape[0]):
+            _khat[i, j] += (
+              1./ (pi * sqrt( 1 - x[i, j] ** 2)) * (
+                exp(-1. / (2 * (1 - x[i,j] ** 2)) * (j_ind[k, j] ** 2 + k_ind[l, j] ** 2 - 2 * x[i, j] * j_ind[k, j] * k_ind[l,j]))
+                + exp(-1. / (2 * (1 - x[i,j] ** 2)) * (j_ind[k, j] ** 2 + k_ind[l, j] ** 2 + 2 * x[i, j] * j_ind[k, j] * k_ind[l,j]))
+              )
+            )
+
+  return khat
+
+
+cpdef get_khat(rho, sig1, sig2):
+    rho = np.ascontiguousarray(rho, dtype=np.float64)
+    squeeze = False
+    if rho.ndim == 1:
+      squeeze = True
+      rho = rho.reshape(1, -1)
+    sig1 =  np.ascontiguousarray(sig1, dtype=np.float64)
+    sig2 =  np.ascontiguousarray(sig2, dtype=np.float64)
+
+    khat = np.zeros((rho.shape[0], rho.shape[1]), dtype=np.float64)
+    khat = _compute_khat(
+        rho, sig1, sig2
+    )
+
+    if squeeze:
+      khat = khat.squeeze()
+    return khat
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef numpy.ndarray[ndim=2, dtype=numpy.float64_t[:, ::1]] _get_cheby_coeff(
+    numpy.ndarray[ndim=3, dtype=numpy.float64_t] rho_coeff,
+    numpy.ndarray[ndim=1, dtype=numpy.int64_t] sv_inds_right1,
+    numpy.ndarray[ndim=1, dtype=numpy.int64_t] sv_inds_right2,
+    numpy.ndarray[ndim=1, dtype=numpy.float64_t] ds1,
+    numpy.ndarray[ndim=1, dtype=numpy.float64_t] ds2
+):
+  cdef int i
+  cdef int j
+  cdef int n = len(ds1)
+  cdef numpy.ndarray[ndim=2, dtype=numpy.float64_t] t = np.zeros((n, 3), dtype=np.float64)
+
+  for i in cython.parallel.prange(n, nogil=True):
```

Comment:
I am a little weary about the blind use of prange here. It might be worth while doing a parallelism trade off study in an ipython notebook to see if there is a breakpoint for the size of `n` where this becomes good. The reason you can't just use prange is because of the overhead in python involved with releasing the gil and making the openmpi calls.

Also i'm curious how this behaves on macosx. Since python is usually build against clang does prange just behave like range?

Reply:
That makes sense. The smallest `n` that can go into this function is the number of baselines, which for the MWA is 8256. That is also a rare case, such as the tiny test file. General use cases will have `n` at least ~2 orders of magnitude greater.

Reply:
Ah interesting. probably big enough it won't matter :thinking: Thanks  for the quick reply.

Reply:
no problem!